compiler/rustc_const_eval/src/const_eval/mod.rs

```diff
@@ -90,19 +115,42 @@ fn const_to_valtree_inner<'tcx>(
         // Technically we could allow function pointers (represented as `ty::Instance`), but this is not guaranteed to
         // agree with runtime equality tests.
         ty::FnPtr(_) | ty::RawPtr(_) => None,
-        ty::Ref(..) => unimplemented!("need to use deref_const"),
+
+        ty::Ref(_, inner_ty, _)  => {
+            match inner_ty.kind() {
+                ty::Slice(_) | ty::Str => {
```

Comment:
This is odd, why are these types treated differently than everything else?

Reply:
We do need to encode the length of the slice in the valtree, but don't need that for non-slice Refs. This is necessary for the Valtree -> ConstValue/ConstAlloc (whatever we end up with) conversion.

Reply:
What about types like `&(i32, [i32])`?

I think you might need something general for the *metadata* of a wide reference, but
- not just slices/str have metadata
- that shouldn't affect the way the pointee is handled recursively

Reply:
Thinking about it some more -- why do you even need to encode the length? It seems redundant. For `&[1,2,3]` at type `&[i32]`, I expect the valtree would be something like
```
   (root)
  /  |  \
1    2   3
```
This contains all the information needed to reconstruct the original value.

Reply:
@RalfJung You're right, we can just read off the length of the slice from the length of the branch slice. Thanks for noticing that we don't actually need to encode the length in the valtree. But I still believe that we need to distinguish between refs corresponding to slices and other refs, because of the indexing into the place (we need to use `mplace_index` for slices and `mplace_field` for everything else), though we can do that in `branches` if you prefer that.

Reply:
I think what Ralf is trying to say is that this will also work if you move the code below into the `Str` and `Slice` arms in the main match and just unconditionally execute the logic in this match's `_` arm.

Reply:
Yeah, the big match in const_to_valtree_inner should get a `ty::Slice` and `ty::Str` arm.